File: NestJs/src/features/auth/auth.service.ts
import { Injectable, BadRequestException, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { UsersRepository } from '@features/users/repositories/users.repository';
import { MailService } from '@infrastructure/mail/mail.service';
import { CreateUserDto, LoginDto } from './dtos/auth.dtos';
import { v4 as uuidv4 } from 'uuid';

@Injectable()
export class AuthService {
  constructor(
    private usersRepository: UsersRepository,
    private jwtService: JwtService,
    private mailService: MailService,
  ) {}

  async register(createUserDto: CreateUserDto) {
    const { email, password, name } = createUserDto;
    const userExists = await this.usersRepository.emailExists(email);

    if (userExists) {
      throw new BadRequestException('Email já cadastrado');
    }
    const hashedPassword = await bcrypt.hash(password, 10);
    const emailConfirmationToken = uuidv4();
    const emailConfirmationExpires = new Date(Date.now() + 24 * 60 * 60 * 1000); // 24 horas
    const user = await this.usersRepository.transaction(async (tx) => {
      const newUser = await this.usersRepository.createWithTransaction(tx, {
        email,
        name,
        password: hashedPassword,
        emailConfirmationToken,
        emailConfirmationExpires,
        emailConfirmed: false,
      });
      await this.mailService.sendConfirmationEmail(email, emailConfirmationToken);

      return newUser;
    });

    return {
      message: 'Usuário criado com sucesso. Verifique seu email para confirmar a conta.',
      user: {
        id: user.id,
        email: user.email,
        name: user.name,
      },
    };
  }

  async confirmEmail(token: string) {
    const user = await this.usersRepository.findByEmailConfirmationToken(token);

    if (!user) {
      throw new BadRequestException('Token de confirmação inválido ou expirado');
    }

    await this.usersRepository.confirmEmail(user.id);

    return {
      message: 'Email confirmado com sucesso!',
    };
  }

  async login(loginDto: LoginDto) {
    const { email, password } = loginDto;

    const user = await this.usersRepository.findByEmail(email);

    if (!user) {
      throw new UnauthorizedException('Email ou senha incorretos');
    }
    if (!user.emailConfirmed) {
      throw new BadRequestException('Por favor, confirme seu email para fazer login');
    }
    const passwordMatches = await bcrypt.compare(password, user.password);

    if (!passwordMatches) {
      throw new UnauthorizedException('Email ou senha incorretos');
    }

    const accessToken = this.jwtService.sign({
      sub: user.id,
      email: user.email,
    });

    return {
      accessToken,
      user: {
        id: user.id,
        email: user.email,
        name: user.name,
      },
    };
  }

  async validateUser(userId: string) {
    const user = await this.usersRepository.findById(userId);

    if (!user) {
      throw new UnauthorizedException('Usuário não encontrado');
    }

    return user;
  }
}

File: SpringBoot/src/main/java/com/demo/features/auth/service/AuthService.java
package com.demo.features.auth.service;

import com.demo.common.exception.BusinessException;
import com.demo.domain.User;
import com.demo.features.auth.dto.AuthResponse;
import com.demo.features.auth.dto.LoginRequest;
import com.demo.features.auth.dto.MessageResponse;
import com.demo.features.auth.dto.RegisterRequest;
import com.demo.features.auth.dto.RegisterResponse;
import com.demo.features.users.repository.UserRepository;
import com.demo.infrastructure.mail.IMailService;
import com.demo.infrastructure.security.JwtService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final IMailService mailService;
    private final ApplicationEventPublisher eventPublisher;

    @org.springframework.beans.factory.annotation.Value("${app.auth.require-email-confirmation:true}")
    private boolean requireEmailConfirmation;

    @Transactional
    public RegisterResponse register(RegisterRequest request) {
        log.info("Registering new user: {}", request.email());

        if (userRepository.existsByEmail(request.email())) {
            throw new BusinessException("Email já está em uso");
        }

        String emailConfirmationToken = UUID.randomUUID().toString();
        LocalDateTime tokenExpires = LocalDateTime.now().plusHours(24);

        User user = User.builder()
                .name(request.name())
                .email(request.email())
                .password(passwordEncoder.encode(request.password()))
                .role("USER")
                .emailConfirmed(false)
                .emailConfirmationToken(emailConfirmationToken)
                .emailConfirmationExpires(tokenExpires)
                .build();

        user = userRepository.save(user);

        // Enviar email de forma assíncrona após o commit
        String userEmail = user.getEmail();
        String token = emailConfirmationToken;

        // Usando TransactionSynchronization para garantir que o email é enviado após o commit
        org.springframework.transaction.support.TransactionSynchronizationManager.registerSynchronization(
            new org.springframework.transaction.support.TransactionSynchronization() {
                @Override
                public void afterCommit() {
                    try {
                        mailService.sendConfirmationEmail(userEmail, token);
                        log.info("✅ Confirmation email queued for: {}", userEmail);
                    } catch (Exception e) {
                        log.error("❌ Failed to queue confirmation email: {}", e.getMessage());
                    }
                }
            }
        );

        log.info("User registered successfully: {}", user.getId());

        return new RegisterResponse(
                "Usuário registrado com sucesso! Verifique seu email para confirmar a conta.",
                new RegisterResponse.UserInfo(user.getId(), user.getEmail(), user.getName())
        );
    }

    @Transactional
    public MessageResponse confirmEmail(String token) {
        log.info("Confirming email with token: {}", token);

        User user = userRepository.findByEmailConfirmationToken(token)
                .orElseThrow(() -> new BusinessException("Token de confirmação inválido"));

        if (user.getEmailConfirmed()) {
            return new MessageResponse("Email já foi confirmado anteriormente");
        }

        if (user.getEmailConfirmationExpires().isBefore(LocalDateTime.now())) {
            throw new BusinessException("Token de confirmação expirado");
        }

        user.setEmailConfirmed(true);
        user.setEmailConfirmationToken(null);
        user.setEmailConfirmationExpires(null);
        userRepository.save(user);

        log.info("Email confirmed for user: {}", user.getId());

        return new MessageResponse("Email confirmado com sucesso! Agora você pode fazer login.");
    }

    public AuthResponse login(LoginRequest request) {
        log.info("User login attempt: {}", request.email());

        User user = userRepository.findByEmail(request.email())
                .orElseThrow(() -> new BusinessException("Email ou senha inválidos"));

        if (!passwordEncoder.matches(request.password(), user.getPassword())) {
            throw new BusinessException("Email ou senha inválidos");
        }

        if (requireEmailConfirmation && !user.getEmailConfirmed()) {
            throw new BusinessException("Email não confirmado. Verifique sua caixa de entrada.");
        }

        List<String> roles = List.of(user.getRole());
        String token = jwtService.generateToken(user.getId(), user.getEmail(), roles);

        log.info("User logged in successfully: {}", user.getId());

        return new AuthResponse(
                token,
                new AuthResponse.UserInfo(user.getId(), user.getEmail(), user.getName())
        );
    }
}
